src/main/java/com/limechain/trie/TrieAccessor.java

```diff
@@ -172,7 +150,31 @@ public void persistUpdates() {
 
     @Override
     public Optional<Nibbles> getNextKey(Nibbles key) {
-        return Optional.ofNullable(trieStorage.getNextKeyByMerkleValue(lastRoot, key));
+        NodeHandle<NodeData> rootHandle = initialTrie.getRootNode().orElse(null);
+        return findNextKey(rootHandle, key, Nibbles.EMPTY);
+    }
+
+    private Optional<Nibbles> findNextKey(NodeHandle<NodeData> node, Nibbles prefix, Nibbles currentPath) {
```

Comment:
Just a question: due to issues with in-memory diffs mismatching the on-disk stored state, this implementations moves the entire logic of finding the next key onto the mutated in-memory trie, right?

Reply:
with the current implementation, yes